src/System.Private.CoreLib/src/System/MulticastDelegate.cs

```diff
@@ -302,7 +303,7 @@ protected override sealed Delegate CombineImpl(Delegate follow)
 
         private object[] DeleteFromInvocationList(object[] invocationList, int invocationCount, int deleteIndex, int deleteCount)
         {
-            object[] thisInvocationList = _invocationList as object[];
+            object[] thisInvocationList = (_invocationList as object[])!; // Calling code ensures _invocationList is object[]
```

Comment:
Another case where I think it'd be better to do something like:
```C#
Debug.Assert(_invocationList is object[], $"Expected {typeof(object[])}, got {_invocationList}");
object[] thisInvocationList = (object[])_invocationList;
```
The assert should prevent the need to use `!` (if it doesn't that should be considered a compiler bug), and if the cast would fail, we'd null ref on the next line; this way at least if we have the wrong type we'll throw a more valuable cast exception.

Reply:
Without the assert, you would get 
```
Unhandled Exception: System.InvalidCastException: Unable to cast object of type 'System.Object' to type 'System.Object[]`
```

With the assert, you would get something like
```
UnhandledExceptiion: Expected System.Object[], got System.Object
```

That doesn't seem an improvement?

Reply:
In terms of code readability, the assert may be slightly better than the comment. I'll add it.

Reply:
> Without the assert, you would get

The assert would fail fast; the cast exception will propagate and could be caught and eaten and impact other logic.

Reply:
Oh - I think I see your point. I was comparing to `object[] thisInvocationList = ((object[])_invocationList)!;`